New `otelcol.exporter.debug` component

BarunKGP opened the pull request:
<!--

CONTRIBUTORS GUIDE: https://github.com/grafana/agent/blob/main/docs/developer/contributing.md#updating-the-changelog

If this is your first PR or you have not contributed in a while, we recommend
taking the time to review the guide. It gives helpful instructions for
contributors around things like how to update the changelog.

-->

#### PR Description
Added a new `otelcol.exporter.debug` component based on the [otelcol debugexporter](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/debugexporter).

#### Which issue(s) this PR fixes

<!-- Uncomment the following line if you want that GitHub issue gets automatically closed after merging the PR -->
Fixes grafana/alloy#301 

#### Notes to the Reviewer
- Based on same design as other `otelcol.exporter` components, specifically the `otelcol.exporter.logging` component (as debug exporter is replacing logging exporter).
- Uses the same defaults as specified in the [otelcol debugexporter factory](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/debugexporter/factory.go)
- Added `exporterArguments` type which is a wrapper over `Arguments` and maintains an implementation that is consistent with `debugexporter`.
  - Removed `DebugMetrics` field from `Arguments` struct. However, it couldn't be completely removed because the upstream `exporter.Argument` specifies this in its interface.
  - `Arguments.Verbosity` is now a string and can be set to a string (such as `"detailed"`). `exporterArguments` deals with the conversion into `configtelemetry.Level` type
- Added tests to check that component `Config` is created from `Arguments` and error is thrown when an invalid `Verbosity` is passed

#### PR Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] CHANGELOG.md updated
- [x] Tests added

ptodev commented on component/otelcol/exporter/debug/debug.go:
```diff
@@ -0,0 +1,72 @@
+
+// Arguments configures the otelcol.exporter.debug component.
+type Arguments struct {
+	Verbosity          configtelemetry.Level `river:"verbosity,attr,optional"`
```
It'd be good to replace `configtelemetry.Level` with a new `Level` type within the Agent's `component/otelcol/exporter/debug/debug` package. Then it can be translated to the OpenTelemetry type `configtelemetry.Level` using a Convert() function. This has a few benefits:
* We would be able to set `verbosity` via a string (e.g. `verbosity = "detailed"`). I am not sure if we can do it now.
* We get a bit more control of what values can be set in the Agent's config.

ptodev commented on component/otelcol/exporter/debug/debug.go:
```diff
@@ -0,0 +1,72 @@
+	SamplingInitial    int                   `river:"sampling_initial,attr,optional"`
+	SamplingThereafter int                   `river:"sampling_thereafter,attr,optional"`
+
+	// DebugMetrics configures component internal metrics. Optional.
+	DebugMetrics otelcol.DebugMetricsArguments `river:"debug_metrics,block,optional"`
```
I think we don't actually need `debug_metrics` in  `otelcol.exporter.debug`. It's ok to remove it. It is only really useful for exporter/receiver which send/receive data over the network. 

`otelcol.exporter.logging` has it, but I think it was unnecessary that we'd added it there. We can leave it in `otelcol.exporter.logging`, but it would help keep `otelcol.exporter.debug` cleaner and less confusing if we don't have it.

ptodev commented on go.mod:
```diff
@@ -616,6 +616,7 @@ require (
 	github.com/natefinch/atomic v1.0.1
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/filterprocessor v0.87.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver v0.87.0
+	go.opentelemetry.io/collector/exporter/debugexporter v0.89.0
```
There are lots of changes to the go.mod. Updating `go.opentelemetry.io/collector` could impact other components, so we normally update such OTel dependencies in one big batch whenever we need to upgrade to a new version.

I think the reason lots of modules got updated is that `debugexporter` is imported as v0.89.0 instead of v0.87.0. Would you mind resetting the changes to go.mod, and instead just adding this line manually:
```
go.opentelemetry.io/collector/exporter/debugexporter v0.87.0
```
In theory that should be the only necessary change to go.mod.

ptodev reviewed (changes requested):
Thank you for your contribution! I left a few comments. Would you mind also adding a couple of things please:
* Unit tests - they need to be in a debug_test.go file, similar to [spanmetrics_test.go](https://github.com/grafana/agent/blob/main/component/otelcol/connector/spanmetrics/spanmetrics_test.go).
* A documentation page - it could be similar to [otelcol.exporter.logging.md](https://github.com/grafana/agent/blob/main/docs/sources/flow/reference/components/otelcol.exporter.logging.md)

ptodev commented on go.sum:
There are still some leftover unnecessary changes in `go.sum`. Would you mind doing a `git reset` on `go.sum`, running `go mod tidy` and rebuilding the Agent please? It shoul help cleanup go.sum.

BarunKGP replied:
Done

ptodev commented on CHANGELOG.md:
```diff
@@ -20,6 +20,8 @@ Main (unreleased)
 
 ### Enhancements
 
+- Added `otelcol.exporter.debug` component based on `go.opentelemetry.io/collector/exporter/debugexporter`. (@BarunKGP)
```
```suggestion
- Added a new `otelcol.exporter.debug` component based on OTel Collector's "debug" exporter`. (@BarunKGP)
```

ptodev reviewed (commented):
Hi, thank you for the update! I don't think it's necessary to have a `exporterArguments` though. It makes adding new arguments more inconvenient, since then we have to add them to both `exporterArguments` and `Arguments` it also makes it less clear how default value are applied, and could make the component a bit more slow due to extra copying.

> I'll add it within the next couple of days, if that's okay?

Yes, of course, thank you!

ptodev commented on component/otelcol/exporter/debug/debug.go:
```diff
@@ -0,0 +1,108 @@
+
+// Arguments configures the otelcol.exporter.debug component.
+type exporterArguments struct {
+	Verbosity          configtelemetry.Level `river:"verbosity, attr, optional"`
+	SamplingInitial    int                   `river:"sampling_initial,attr, optional"`
+	SamplingThereafter int                   `river:"sampling_thereafter,attr, optional"`
```
TBH I'd just remove `exporterArguments`, I don't think it's really necessary. Converting from `Arguments` to the OTel type is ok.

BarunKGP replied:
Removed `exporterArguments`

BarunKGP commented on component/otelcol/exporter/debug/debug_test.go:
```diff
@@ -0,0 +1,33 @@
+	otelcomponent "go.opentelemetry.io/collector/component"
+)
+
+func Test(t *testing.T) {
```
Added tests in table format similar to ones in `spanmetrics_test.go`

BarunKGP commented:
@ptodev happy new year! Apologies for the delay in making the requested changes, I got caught up in other stuff over the holidays.

I've now addressed all the review comments you left. In summary:
- Reset `go.sum` and reverted to the previous dependency versions
- Added tests to `debug_test.go` in table format and checked that they are passing
- Removed unnecessary `exporterArguments` field 
- Edited changelog for clarity

Please let me know if there are any other changes you'd like me to make. Thanks!

ptodev commented:
Hi @BarunKGP! Happy new year, and apologies for the late reply! The code looks great, thank you!

The only remaining task is to add a docs page similar to [otelcol.exporter.logging.md](https://github.com/grafana/agent/blob/main/docs/sources/flow/reference/components/otelcol.exporter.logging.md). Would you be able to add it please?

BarunKGP commented:
@ptodev I've created the doc, but what about the aliases section as in [otelcol.exporter.logging.md](https://github.com/grafana/agent/blob/main/docs/sources/flow/reference/components/otelcol.exporter.logging.md)? Should I remove it, or do we have any new aliases to include here?

Also taking a look at the linter issue. I'll push my changes by this weekend.

ptodev commented:
> @ptodev I've created the doc, but what about the aliases section as in [otelcol.exporter.logging.md](https://github.com/grafana/agent/blob/main/docs/sources/flow/reference/components/otelcol.exporter.logging.md)? Should I remove it, or do we have any new aliases to include here?

Thank you, the docs look good! It is best to retain the same aliases as `otelcol.exporter.logging`, so the doc is good as it is.

> Also taking a look at the linter issue. I'll push my changes by this weekend.

Thank you! I'll give it a final stamp of approval afterwards. It would be nice to rebase the branch with `main` too.

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.exporter.debug.md:
```diff
@@ -0,0 +1,88 @@
+
+# otelcol.exporter.debug
+
+`otelcol.exporter.debug` accepts telemetry data from other `otelcol` components
+and writes them to the console (stderr). The verbosity of the logs can also be controlled.
```
```suggestion
`otelcol.exporter.debug` accepts telemetry data from other `otelcol` components and writes them to the console (stderr).
You can control the verbosity of the logs.
```

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.exporter.debug.md:
```diff
@@ -0,0 +1,88 @@
+`otelcol.exporter.debug` accepts telemetry data from other `otelcol` components
+and writes them to the console (stderr). The verbosity of the logs can also be controlled.
+
+> **NOTE**: `otelcol.exporter.debug` is a wrapper over the upstream
+> OpenTelemetry Collector `debug` exporter. Bug reports or feature requests will
+> be redirected to the upstream repository, if necessary.
```
```suggestion
{{< admonition type="note" >}}
otelcol.exporter.debug` is a wrapper over the upstream OpenTelemetry Collector `debug` exporter.
If necessary, bug reports or feature requests are redirected to the upstream repository.
{{< /admonition >}}
```

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.exporter.debug.md:
```diff
@@ -0,0 +1,88 @@
+> OpenTelemetry Collector `debug` exporter. Bug reports or feature requests will
+> be redirected to the upstream repository, if necessary.
+
+Multiple `otelcol.exporter.debug` components can be specified by giving them
+different labels.
```
```suggestion
You can specify multiple `otelcol.exporter.debug` components by giving them different labels.
```

We can break on sentences vs arbitrary (80 column).

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.exporter.debug.md:
```diff
@@ -0,0 +1,88 @@
+
+## Example
+
+This example scrapes prometheus unix metrics and writes them to the console:
```
```suggestion
This example scrapes Prometheus UNIX metrics and writes them to the console:
```

clayton-cornell reviewed (commented):
Some doc input.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/agent?pullRequest=5867) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/agent?pullRequest=5867) before we can accept your contribution.<br/>**17** out of **18** committers have signed the CLA.<br/><br/>:white_check_mark: korniltsev<br/>:white_check_mark: jcreixell<br/>:white_check_mark: marctc<br/>:white_check_mark: wildum<br/>:white_check_mark: bboreham<br/>:white_check_mark: mattdurham<br/>:white_check_mark: clayton-cornell<br/>:white_check_mark: hainenber<br/>:white_check_mark: captncraig<br/>:white_check_mark: rfratto<br/>:white_check_mark: tpaschalis<br/>:white_check_mark: ptodev<br/>:white_check_mark: erikbaranowski<br/>:white_check_mark: StefanKurek<br/>:white_check_mark: kurczynski<br/>:white_check_mark: simonswine<br/>:white_check_mark: BarunKGP<br/>:x: github-actions[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/agent?pullRequest=5867) it.</sub>

BarunKGP commented:
@clayton-cornell  Thank you for the doc suggestions. I have accepted and incorporated all of them.

@ptodev rebased the branch. Let me know if there's anything else I should take a look at :)

ptodev commented:
@BarunKGP - The rebase must have gone wrong, because there are lots of changes in the PR now.

BarunKGP commented:
@ptodev I've reverted this branch to its state prior to the rebase and added the doc changes to it. Upon investigating, it looks like I was incorporating suggestions from these comments onto my origin/main directly, which is why the two branches diverged.

I'm thinking of creating a new branch with all my changes from here, while I force pull `upstream/main` on this one and then rebase my changes on top of it. Unfortunately, the git log will get a bit messy. Do you have any suggestions on how to go about this?

ptodev commented:
This is the way I normally rebase my feature branches:
```
git fetch
git rebase origin/main
```

If you'd like to clean up your branch in a more step-by-step basis, you can also do something like:
```
# Remove your commits but retain your changes
git reset --soft <comit-before-your-commits>`

git stash

# Make your branch the same as main
# I'm not too sure about this one, but I think it'd work?
git reset --hard main

git stash pop
```

You could also just create a new branch off of main, apply your changes there, and then use `git reset` on your old branch to make it point to the same commit as the new one.

BarunKGP commented:
I've added all the changes to the branch after rebasing. I've also updated the new components to the revised API for `components`.

What should be the `Stability` for this however? I've left it as `StabilityExperimental` for now.

ptodev commented on internal/component/otelcol/exporter/debug/debug.go:
```diff
@@ -0,0 +1,96 @@
+
+// DefaultArguments holds default values for Arguments.
+var DefaultArguments = Arguments{
+	Verbosity:          "basic",
```
```suggestion
	Verbosity:          "normal",
```

ptodev commented on docs/sources/flow/reference/components/otelcol.exporter.debug.md:
```diff
@@ -0,0 +1,88 @@
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`verbosity`           | `string` | Verbosity of the generated logs. | `"basic"` | no
```
```suggestion
`verbosity`           | `string` | Verbosity of the generated logs. | `"normal"` | no
```
The default [upstream](https://github.com/open-telemetry/opentelemetry-collector/tree/v0.86.0/exporter/debugexporter) is "normal", so let's use the same one here.

ptodev commented on internal/component/otelcol/exporter/debug/debug_test.go:
```diff
@@ -0,0 +1,83 @@
+		{
+			testName: "defaultConfig",
+			args: `
+			verbosity = "basic"
+			sampling_initial = 2
+			sampling_thereafter = 500
```
```suggestion
```

In order to truly test if the defaults are set correctly, we need to not supply any arguments.

ptodev commented on go.mod:
```diff
@@ -620,6 +620,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/filterprocessor v0.87.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver v0.87.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver v0.87.0
+	go.opentelemetry.io/collector/exporter/debugexporter v0.87.0
```
Are all the changes to the go.mod really necessary? This line is the only one I'd expect us to add.

ptodev reviewed (commented):
Thank you for the update!

> What should be the `Stability` for this however? I've left it as `StabilityExperimental` for now.

"Experimental" is the right one, yes.

There is a [linter error](https://drone.grafana.net/grafana/agent/17377/1/2):
```
internal/component/otelcol/exporter/debug/debug.go:9: File is not `gofmt`-ed with `-s` (gofmt)
```

I also added a few minor comments. After they are resolved, I think it should be ok for merging.

BarunKGP commented:
@ptodev thank you for your time in reviewing the PR! 

I have accepted all suggestions and made the following changes addressing them:

- Fixed imports in go.mod -> only added debugexporter as a dependency
- Updated default arguments and test for default arguments
- Fixed linting issues

Please let me know if there are any other areas I need to work on  😃

ptodev reviewed (approved):
Thank you very much! LGTM 🚀

clayton-cornell commented:
@ptodev This one seems to have been missed... is it still valid? Does anything been to be moved to Alloy?

clayton-cornell reviewed (approved):
Docs are OK as-is

clayton-cornell commented:
@ptodev There is a linting error (outside the docs) that I can't fix.  Any ideas?

ptodev commented:
@clayton-cornell I fixed it, no worries. Thank you for the review!
@BarunKGP Thank you for your contribution, and apologies for the delay with the review!

I also opened a [PR](https://github.com/grafana/alloy/pull/1043) to add the component to Alloy.
